src/blog/memory-safe-languages/myths-zero-overhead-memory-safety.vmd

```diff
@@ -0,0 +1,308 @@
+ * Hash maps could be expensive depending on the hashing algorithms, such as ones that use modulus.
+ * Generational references access the object's field in parallel with its generation which is likely on the same cache line, but it can be expensive in tight loops unless one uses [linear style](https://verdagon.dev/blog/linear-types-borrowing).
+
+
+To really get good performance, one has to *pick their poison* and decide what overhead they're willing to suffer.
+
+
+Though if we're lucky, a particular situation has something special that makes some approaches less costly than others, which is always nice.
+
+
+<ignore>
+#factorio: The game [Factorio](https://www.factorio.com/) is a marvel of modern optimization. Someone once asked,
+
+> How can you prevent "random access" to memory when those inserters and assemblers are stored in different lists and presumably in different indexes in those lists?
+
+One of the Factorio developers [replied](https://www.reddit.com/r/factorio/comments/13bsf3s/comment/jjcra0t/),
+
+> You can't; it's the nature of complex systems; they end up interacting with other systems. In the end you just have to pay the cost to access that memory if you want the inserters to do what inserters do.
+</ignore>
+
+
+<slice />
+
+# What does it all mean?
+
+There's no way to get to zero-overhead memory safety.
+
+However, *we can get close.*
+
+
+For language designers, it means that the search isn't over. Since there is still overhead, there is still room to improve or rearrange.
+
+Perhaps it means that we should *keep chasing the myth* of zero-overhead memory safety, because we aren't there yet.
+
+That's what I'm trying to do with Vale, and led to discovering its [linear-aliasing model](https://vale.dev/linear-aliasing-model), generational references, and [regions](/blog/zero-cost-borrowing-regions-overview).
+
+A lot of other languages are exploring these areas too, like [Austral](https://austral-lang.org/), [Val](https://val-lang.dev/), and [Verona](https://github.com/microsoft/verona).
+
+
+Perhaps it means that we can *stop chasing the myth*.
+
+Back when I coded in Rust a lot more, I made the mistake of caring too much about avoiding heap allocations, `Rc`, `RefCell`, virtual calls, etc. It was idiomatic to avoid these particular sources of overhead. I took a sense of pride in going the extra mile and spending the extra time reducing my overhead down to "zero".
+
+In retrospect, I never actually reached zero, because instead I used more cloning, more bounds checking, and more hashing. It was an improvement, but I was chasing a definition of perfection that was arbitrary and unhelpful. I wasted a lot of time past the point of diminishing returns, chasing that myth.
+
+It wasn't a problem with Rust. It was a problem with me. Perhaps instead of taking a community's idioms as gospel, we should spend that time doing more profiling.
```

Comment:
This is a pretty common sentiment, and where the term "premature optimization" comes from.